crates/bevy_render/src/render_resource/shader.rs

```diff
@@ -401,16 +409,62 @@ impl ShaderProcessor {
             }
         };
 
-        let shader_defs_unique = HashSet::<String>::from_iter(shader_defs.iter().cloned());
+        let shader_defs_unique =
+            HashMap::<String, ShaderDefVal>::from_iter(shader_defs.iter().map(|v| match v {
+                ShaderDefVal::Bool(k, _) | ShaderDefVal::Int(k, _) => (k.clone(), v.clone()),
+            }));
         let mut scopes = vec![true];
         let mut final_string = String::new();
         for line in shader_str.lines() {
             if let Some(cap) = self.ifdef_regex.captures(line) {
                 let def = cap.get(1).unwrap();
-                scopes.push(*scopes.last().unwrap() && shader_defs_unique.contains(def.as_str()));
+                scopes
+                    .push(*scopes.last().unwrap() && shader_defs_unique.contains_key(def.as_str()));
             } else if let Some(cap) = self.ifndef_regex.captures(line) {
                 let def = cap.get(1).unwrap();
-                scopes.push(*scopes.last().unwrap() && !shader_defs_unique.contains(def.as_str()));
+                scopes.push(
+                    *scopes.last().unwrap() && !shader_defs_unique.contains_key(def.as_str()),
+                );
+            } else if let Some(cap) = self.ifeq_regex.captures(line) {
+                let def = cap.get(1).unwrap();
+                let val = cap.get(2).unwrap();
+                if let Some(def) = shader_defs_unique.get(def.as_str()) {
+                    let new_scope = match def {
+                        ShaderDefVal::Bool(_, def) => val
+                            .as_str()
+                            .parse()
+                            .map(|val: bool| *def == val)
+                            .unwrap_or_default(),
+                        ShaderDefVal::Int(_, def) => val
+                            .as_str()
+                            .parse()
+                            .map(|val: i32| *def == val)
+                            .unwrap_or_default(),
+                    };
+                    scopes.push(*scopes.last().unwrap() && new_scope);
+                } else {
+                    scopes.push(false);
```

Comment:
For expressions on shader def values we might want to have hard error messages if the shader def does not exist. Implicitly treating if expressions containing undefined defs as `false` could be weird:

```rust
// This reads as "if FEATURE is supported, do X"
#if FEATURE_IS_NOT_SUPPORTED == false
DO_X
#endif

// if FEATURE_IS_NOT_SUPPORTED is not defined, implicitly the feature is supported
// (or at least, it reads this way)
// but the expression will still be evaluated as false under the current rules
```